internal/gitutil/gitutil.go

```diff
@@ -359,28 +359,40 @@ func (gur *GitUpstreamRepo) cacheRepo(ctx context.Context, uri string, requiredR
 		gitRunner.Dir = repoCacheDir
 	}
 
-	// fetch the specified refs
-	triedFallback := false
+loop:
 	for _, s := range requiredRefs {
-		if _, err := gitRunner.Run(ctx, "fetch", "origin", "--depth=1", s); err != nil {
-			if !triedFallback { // only fallback to fetch origin once
-				// fallback on fetching the origin. If the user provided a short sha,
-				// we need to fetch all objects in order to resolve it into the full
-				// sha.
-				// TODO: See if there is a way to resolve a short sha into a complete
-				// sha without fetching. Haven't found one so far...
-				if _, retryErr := gitRunner.Run(ctx, "fetch", "origin"); retryErr != nil {
-					// We are using the original error here.
-					return "", errors.E(op, errors.Git, fmt.Errorf(
-						"error running `git fetch` for origin: %w", err))
-				}
-				triedFallback = true
+		// Check if we can verify the ref. This will output a full commit sha if
+		// either the ref (short commit, tag, branch) can be resolved to a full
+		// commit sha, or if the provided ref is already a valid full commit sha (note
+		// that this will happen even if the commit doesn't exist in the local repo).
+		rr, _ := gitRunner.Run(ctx, "rev-parse", "--verify", "-q", s)
```

Comment:
Why are we not checking error here ? What if err != nil and rr = nil ?

Reply:
An error here just means that the ref wasn't found. Instead of using the error, we rely on the fact that with the `-q` flag, `git rev-parse` will output the commit SHA if the ref was found, and nothing if it was not found.
I added an extra line in the comment to explain this.
`rr` can't be nil since it is not a pointer, so `rr.Stdout` always returns a string.

Reply:
Got it. Do you think we need to add a test to compare the stdout ? All the tests we perform on get just checks for repo.

Reply:
I don't think that is needed. The git documentation is pretty clear and we have tests that covers the result of making this check.